projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,246 @@
+    coordinates << "image name" << "," << "timestamp" <<"\n";
+    coordinates.close();
+
+
+    BoBRobotics::BackgroundExceptionCatcher catcher;
+    catcher.trapSignals();
+
+    int i = 0;
+    BoBRobotics::Stopwatch sw, sw_timestamp;
+    std::array<degree_t, 3> angles;
+    // for x timestep
+    sw_timestamp.start();
+    for (;;) {
+
+        try {
+            catcher.check();
+            sw.start();
+            // get imu data
+            angles = imu.getEulerAngles();
+            float yaw = angles[0].value();
+            float pitch = angles[1].value();
+            float roll = angles[2].value();
+
+            // get gps data
+            BoBRobotics::GPS::GPSData data = gps.getGPSData();
+            BoBRobotics::MapCoordinate::GPSCoordinate coord = data.coordinate;
+            BoBRobotics::GPS::TimeStamp time = data.time;
+            int gpsQual = (int) data.gpsQuality; // gps quality
+
+            // output results
+    	   // if (i%5 == 0) { // only output once ever sec
+            std::cout << std::setprecision(10) << "GPS quality: " << gpsQual << " latitude: " << coord.lat.value()<< " longitude: " <<  coord.lon.value()  << " num sats: " << data.numberOfSatellites << " time: " << static_cast<units::time::millisecond_t>(sw_timestamp.elapsed()).value()/1000 << std::endl;
+
+	        std::ostringstream fileString, folderString;
+            fileString << folderName << "/image" << i << ".jpg";
+	        folderString << folderName << "/coordinates.csv";
+            std::string fileName = fileString.str();
+
+            //  headers:   gpsQuality|latitude|longitude|altitude|roll|pitch|yaw|imagename|timestamp
+            coordinates.open (folderString.str(), std::ofstream::app); // open coordinates.csv file and append
+            coordinates << std::setprecision(10) << std::fixed << gpsQual << "," << coord.lat.value() << "," << coord.lon.value() << "," << data.altitude.value()  << "," << roll << "," << pitch << "," << yaw  << ",";
+            coordinates << "image" << i << ".jpg" << "," << static_cast<units::time::millisecond_t>(sw_timestamp.elapsed()).value() <<"\n";
+            coordinates.close();
+
+            // poll from camera thread
+            cv::Mat originalImage;
+            mex.lock();
+            tmpImage.copyTo(originalImage);
+            mex.unlock();
+
+            if (originalImage.size().height > 0) {
+               // cv::resize(originalImage, originalImage, cv::Size(), 0.5, 0.5);
+               // cv::imshow("orig",resized);
+            }
+
+            std::thread imwriteThread{ &saveThread, std::move(originalImage), std::move(fileName) };
+            imwriteThread.detach();
+
+            int count = (std::chrono::milliseconds(UPDATE_INTERVAL) - sw.elapsed()).count();
+            //std::cout << " count " << count/1000000 << " " << std::endl;
+            count /= 1000000;
```

Comment:
why not have count in whatever units you actually want it in?

i.e. 
```c++
std::chrono::milliseconds count = std::chrono::milliseconds(UPDATE_INTERVAL) - sw.elapsed();
std::this_thread::sleep_for(count);
```

Reply:
In a similar vein to @neworderofjamie's suggestion above, you could also do something like:
```c++
constexpr auto UpdateInterval = 200ms;
```

You would then need a ``using namespace std::literals`` somewhere above.